File: ios/RNMBX/RNMBXModels.swift
@_spi(Experimental) import MapboxMaps

@objc(RNMBXModels)
open class RNMBXModels : UIView, RNMBXMapComponent {
  var modelIdToUrl: [String: String] = [:]
  
  @objc(models)
  public var models: NSDictionary {
    set {
      var newModels: [String: String] = [:]
      newValue.forEach { (key, value) in
        if let value = value as? NSDictionary, let key = key as? String {
          if let uri = value["uri"] as? String {
            newModels[key] = uri
          } else if let url = value["url"] as? String {
            newModels[key] = url
          } else {
            Logger.log(level: .error, message: "Unexpected value for model key: \(key) \(value) - no uri or url found")
          }
        }
      }
      self.modelIdToUrl = newModels
    }
    get {
      return modelIdToUrl as NSDictionary
    }
  }
  
  func setModels(_ models: NSDictionary) {

  }
  
  public func addToMap(_ map: RNMBXMapView, style: Style) {
    modelIdToUrl.forEach { (id, uri) in
      logged("Models.addStyleModel") {
        if let link = URL(string: uri), let scheme = link.scheme, let host = link.host,
            let port = link.port {

          // https://github.com/mapbox/mapbox-maps-ios/issues/2067
          let uriWithoutQuery = "\(scheme)://\(host):\(port)\(link.path)"

          try style.addStyleModel(modelId: id, modelUri: uriWithoutQuery)
        } else {
          try style.addStyleModel(modelId: id, modelUri: uri)
        }
      }
    }
  }
  
  public func removeFromMap(_ map: RNMBXMapView, reason: RemovalReason) -> Bool {
    modelIdToUrl.forEach { (id, _) in
      #if RNMBX_11
      try? map._mapView?.mapboxMap.removeStyleModel(modelId: id)
      #endif
    }
    return true
  }
  
  public func waitForStyleLoad() -> Bool {
    true
  }
}

File: android/src/main/java/com/rnmapbox/rnmbx/components/styles/model/RNMBXModels.kt
package com.rnmapbox.rnmbx.components.styles.model

import android.content.Context
import com.facebook.react.bridge.Dynamic
import com.facebook.react.bridge.ReadableMap
import com.rnmapbox.rnmbx.components.AbstractMapFeature
import com.rnmapbox.rnmbx.components.mapview.RNMBXMapView
import java.net.URI

class RNMBXModels(context: Context?) : AbstractMapFeature(context) {
    private var models: Map<String, String>? = null;

    fun setModels(value: Map<String,String>) {
        this.models = value;
        applyModels()
    }

    override fun addToMap(mapView: RNMBXMapView) {
        super.addToMap(mapView)
        applyModels()
    }

    private fun applyModels(){
        mMapView?.mapView?.getMapboxMap()?.getStyle()?.let { style ->
            models?.forEach { (modelId,modelUri) ->
                val uri = URI.create(modelUri)
                val modelUriWithoutQuery = URI(uri.scheme, uri.userInfo, uri.host, uri.port, uri.path, null, uri.fragment).toString()

                style.addStyleModel(modelId, modelUriWithoutQuery)
            }
        }
    }
}